text/0032-minimal-relay.md

```diff
@@ -0,0 +1,196 @@
+# RFC-0032: Minimal Relay
+
+|                 |                                                                               |
+| --------------- | ----------------------------------------------------------------------------- |
+| **Start Date**  | 20 September 2023                                                             |
+| **Description** | Proposal to minimise Relay Chain functionality.                               |
+| **Authors**     | Joe Petrowski, Gavin Wood                                                     |
+
+## Summary
+
+The Relay Chain contains most of the core logic for the Polkadot network. While this was necessary
+prior to the launch of parachains and development of XCM, most of this logic can exist in
+parachains. This is a proposal to migrate several subsystems into system parachains.
+
+## Motivation
+
+Polkadot's scaling approach allows many distinct state machines (known generally as parachains) to
+operate with common guarantees about the validity and security of their state transitions. Polkadot
+provides these common guarantees by executing the state transitions on a strict subset (a backing
+group) of the Relay Chain's validator set.
+
+However, state transitions on the Relay Chain need to be executed by _all_ validators. If any of
+those state transitions can occur on parachains, then the resources of the complement of a single
+backing group could be used to offer more cores. As in, they could be offering more coretime (a.k.a.
+blockspace) to the network.
+
+By minimising state transition logic on the Relay Chain by migrating it into "system chains" -- a
+set of parachains that, with the Relay Chain, make up the Polkadot protocol -- the Polkadot
+Ubiquitous Computer can maximise its primary offering: secure blockspace.
+
+## Stakeholders
+
+- Parachains that interact with affected logic on the Relay Chain;
+- Core protocol and XCM format developers;
+- Tooling, block explorer, and UI developers.
+
+## Explanation
+
+The following pallets and subsystems are good candidates to migrate from the Relay Chain:
+
+- Identity
+- Balances
+- Staking
+	- Staking
+	- Election Provider
+	- Bags List
+	- NIS
+	- Nomination Pools
+	- Fast Unstake
+- Governance
+	- Treasury and Bounties
+	- Conviction Voting
+	- Referenda
+
+Note: The Auctions and Crowdloan pallets will be replaced by Coretime, its system chain and
+interface described in RFC-1 and RFC-5, respectively.
+
+### Migrations
+
+Some subsystems are simpler to move than others. For example, migrating Identity can be done by
+simply preventing state changes in the Relay Chain, using the Identity-related state as the genesis
+for a new chain, and launching that new chain with the genesis and logic (pallet) needed.
+
+Other subsystems cannot experience any downtime like this because they are essential to the
```

Comment:
Co-existing will be quite tricky for staking, and we would likely have to do a one-off migration. The migration is possible to happen within an era, but even if not, having one extended "super era" is not too bad. 

With a one-off migration, all the staking data is moved to the system chain. Some historical staking data will be kept on the RC for retroactive rewards and slashes, and will be removed after a maximum of 84 eras.

Reply:
Is this due to mutual exclusivity of freezing on the Relay-chain and the Staking chain?

Reply:
We discussed an alternative migration strategy to staking chain by running two parallel staking system for a while.
Roughly the way it would work is:

Staking system in the Relay chain works independently of Staking Chain. Relay Chain has a configuration that decides how many validators are provided from staking chain, the rest elected locally in the Relay Chain. We can then slowly scale up this number to 100% Staking Chain.

Validators who want to validate via Staking Chain, would need to unbond, teleport assets to Staking Chain, and set their intent to validate again (we could potentially improve this by allowing teleport of staked assets without unlocking). In the beginning it may be easy to get into the active validator set via Staking chain until we reach some equilibrium eventually getting to similar economic security as Relay Chain. Once we go to 100% from Staking Chain, at some point we can kick all validators/nominators who did not migrate.

The main opposition to this approach were

- Total stake for active validators (economic security) may drop while in transition phase.
- Validators have to move and re-acquire the approval score they gained on the Relay Chain.
- It might be harder to move all nominators voluntarily to Staking Chain.

We (weakly) reached to the consensus that one-off migration might be simpler. We will need to migrate the staking
ledgers and the staked balance to the Staking Chain and do this in one era (which can be an extended era).
The era history can stay in Relay Chain allowing validators to claim historical rewards. Once 84 eras (history depth)
have passed, we purge the historical data and clean up staking code in relay chain.

Keen to hear if you have some thoughts/feedbacks on both these approaches (this probably deserves its own RFC but we do have an [open issue](https://github.com/paritytech/polkadot-sdk/issues/491) where we have some more notes from our discussions so far).

Reply:
This migration will require a lot of upfront communication. The less users and validators have to do, the better. I believe that a one-off migration would be much easier to handle on the wallets side, and communication side. Wallets and partners (think CEXs etc) should be prepared, and when the block hits, they'll be able to switch to displaying/using the staking chain. With a gradual move however, I can see a period when users don't know if they are on the RC or the SC, UIs would have to show and handle both, which would likely be very confusing.

Reply:
I wholeheartedly agree with a one-off migration approach, in both this case and as a general pattern for moving systems between chains or performing a large complex upgrade.

Reply:
We should obviously not require that users change wallet software to make new accounts, transfer, stake, etc.  That's a big no no.

We need domain separation material in signatures of course, but the exact semantics of that domain separation material can evolve if required.  If the new chain is *the* staking chain, then it can have a "virtual" genesis hash of the original relay chain or whatever.  We donno how to have multiple staking chains anyways.

I guess governance could be more nuanced. 

As for migration, we could've the staking chain "fork" off from the relay chain at block height x, which freezes transactions on the relay chain, then the staking chain runs a migration pallet, and finally the staking chain unfreezes its own transactions.  Anyways a chain "fork" operation should not be expensive and should be useful elsewhere.

We'll want a bunch of O(1) drops here, so those should be implemented first.  We might clean up the state a bit first too, like fixing the slashing spans system, and droping that old state.  As a backup, we proved a limited recovery sudo which kills the staking chain and unfreezes the relay chain accounts.

Reply:
> As for migration, we could've the staking chain "fork" off from the relay chain at block height x, which freezes transactions on the relay chain, then the staking chain runs a migration pallet, and finally the staking chain unfreezes its own transactions. Anyways a chain "fork" operation should not be expensive and should be useful elsewhere.

What we really need to move is staking pallet state and part of user balance that is staked. Unstaked balances should eventually end up in AssetHub once we get rid of balances from relay chain (and should be part of separate migration I think). We could do a fork and then get rid of things that we don't need in Staking Chain and translate what we need (I guess thats what your migration pallet would do). Relay chain can only freeze staking related transactions and kill it once staking chain is operational. Does that make sense?

@kianenigma had the idea to prepare data that we need to migrate, calculate the state root and securely send it to staking chain (root track or xcm), and then have a bot that fills the storage on staking chain through permissionless calls. Wdyt? May be forking off is simpler?